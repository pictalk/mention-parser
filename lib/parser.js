"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const safe_types_1 = require("safe-types");
var ErrorCode;
(function (ErrorCode) {
    /**
     * A zero-valued error is not an error.
     * Currently unused.
     */
    ErrorCode[ErrorCode["None"] = 0] = "None";
    /**
     * Syntax error when a tag open/close token is used inside a tag
     * as a type or value.
     */
    ErrorCode[ErrorCode["InvalidToken"] = 1] = "InvalidToken";
    /**
     * Syntax error where the parser is trying to close the tag,
     * but does not receive the correct sequence of tokens.
     */
    ErrorCode[ErrorCode["UnclosedTag"] = 2] = "UnclosedTag";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
function parse(message) {
    const tags = [];
    let state = 0 /* Text */;
    let type = "text" /* Text */;
    let value = "";
    for (let char of Array.from(message)) {
        switch (state) {
            case 0 /* Text */:
                // Switch state on a tag opening token '{'
                if (char === "{" /* Open */) {
                    state = 1 /* Opening */;
                    break;
                }
                value += char;
                break;
            case 1 /* Opening */:
                // Tag open failed--expected double curly braces '{{'
                if (char !== "{" /* Open */) {
                    state = 0 /* Text */;
                    // Return the original '{' back to the value buffer
                    value += "{" /* Open */ + char;
                    break;
                }
                // Tag opened successfully.
                state = 2 /* Type */;
                // Flush the current tag.
                if (value.length) {
                    tags.push({ type, value });
                }
                // New tag
                type = "";
                value = "";
                break;
            case 2 /* Type */:
                // We expect to parse the type, but we can encounter more open tokens
                if (char === "{" /* Open */) {
                    // If we've already buffered some of the type,
                    // this is a syntax error
                    if (type.length > 0) {
                        return safe_types_1.Err({
                            code: ErrorCode.InvalidToken,
                            message: `cannot use tag opening token '${"{" /* Open */}' inside of tag`,
                        });
                    }
                    else {
                        // In this case, we haven't begun buffering the type,
                        // so these open tokens should be parsed as text chars
                        // that belong to the previous text tag.
                        let tail_tag = tags[tags.length - 1];
                        if (!tail_tag || tail_tag.type !== "text" /* Text */) {
                            tags.push({ type: "text" /* Text */, value: "{" /* Open */ });
                            break;
                        }
                        tail_tag.value += "{" /* Open */;
                        break;
                    }
                }
                if (char === "}" /* Close */) {
                    return safe_types_1.Err({
                        code: ErrorCode.InvalidToken,
                        message: `cannot use tag closing token '${"}" /* Close */}' inside of tag`,
                    });
                }
                // Finished consuming tag type, switch to value.
                if (char === ":" /* Sep */) {
                    state = 3 /* Value */;
                    type = type.trim();
                    break;
                }
                type += char;
                break;
            case 3 /* Value */:
                if (char === "{" /* Open */) {
                    return safe_types_1.Err({
                        code: ErrorCode.InvalidToken,
                        message: `cannot use tag opening token '${"{" /* Open */}' inside of tag`,
                    });
                }
                // Finished consuming tag value, switch to close the tag.
                if (char === "}" /* Close */) {
                    state = 4 /* Closing */;
                    value = value.trim();
                    break;
                }
                value += char;
                break;
            case 4 /* Closing */:
                if (char !== "}" /* Close */) {
                    return safe_types_1.Err({
                        code: ErrorCode.UnclosedTag,
                        message: `tag was not properly closed with two '${"}" /* Close */}' tokens`,
                    });
                }
                state = 0 /* Text */;
                tags.push({ type, value });
                // New text tag buffer
                type = "text" /* Text */;
                value = "";
                break;
            default:
                expectNever(state);
        }
    }
    if (state !== 0 /* Text */) {
        return safe_types_1.Err({
            code: ErrorCode.UnclosedTag,
            message: "text parsed without closing a tag",
        });
    }
    // Flush the current tag.
    if (value.length) {
        tags.push({ type, value });
    }
    return safe_types_1.Ok(tags);
}
exports.parse = parse;
function expectNever(_, msg = "unexpected code path") {
    throw new Error(msg);
}
//# sourceMappingURL=parser.js.map