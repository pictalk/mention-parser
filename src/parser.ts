import { Result, Ok, Err } from "safe-types";
import { Tag } from "./shared";

const enum TagType {
  Text = "text",
}

const enum Token {
  Open = "{",
  Close = "}",
  Sep = ":",
}

const enum State {
  Text,
  Opening,
  Type,
  Value,
  Closing,
}

export enum ErrorCode {
  /**
   * A zero-valued error is not an error.
   * Currently unused.
   */
  None,
  /**
   * Syntax error when a tag open/close token is used inside a tag
   * as a type or value.
   */
  InvalidToken,
  /**
   * Syntax error where the parser is trying to close the tag,
   * but does not receive the correct sequence of tokens.
   */
  UnclosedTag,
  /**
   * Type error when a tag is matched that does not exist in a given
   * set of allowed tags.
   */
  UnrecognizedTag,
}

export interface ParseError {
  code: ErrorCode;
  message: string;
}

export function parse<T extends string = string>(
  message: string,
  tagsAllowed?: Set<T>,
): Result<Tag<T>[], ParseError> {
  const tags: Tag[] = [];
  let state = State.Text;
  let type: string = TagType.Text;
  let value = "";

  for (let char of Array.from(message)) {
    switch (state) {
      case State.Text:
        // Switch state on a tag opening token '{'
        if (char === Token.Open) {
          state = State.Opening;
          break;
        }

        value += char;
        break;

      case State.Opening:
        // Tag open failed--expected double curly braces '{{'
        if (char !== Token.Open) {
          state = State.Text;
          // Return the original '{' back to the value buffer
          value += Token.Open + char;
          break;
        }

        // Tag opened successfully.
        state = State.Type;

        // Flush the current tag.
        if (value.length) {
          tags.push({ type, value });
        }

        // New tag
        type = "";
        value = "";
        break;

      case State.Type:
        // We expect to parse the type, but we can encounter more open tokens
        if (char === Token.Open) {
          // If we've already buffered some of the type,
          // this is a syntax error
          if (type.length > 0) {
            return Err({
              code: ErrorCode.InvalidToken,
              message: `cannot use tag opening token '${
                Token.Open
              }' inside of tag`,
            });
          } else {
            // In this case, we haven't begun buffering the type,
            // so these open tokens should be parsed as text chars
            // that belong to the previous text tag.
            let tail_tag = tags[tags.length - 1];
            if (!tail_tag || tail_tag.type !== TagType.Text) {
              tags.push({ type: TagType.Text, value: Token.Open });
              break;
            }

            tail_tag.value += Token.Open;
            break;
          }
        }

        if (char === Token.Close) {
          return Err({
            code: ErrorCode.InvalidToken,
            message: `cannot use tag closing token '${
              Token.Close
            }' inside of tag`,
          });
        }

        // Finished consuming tag type, switch to value.
        if (char === Token.Sep) {
          state = State.Value;
          type = type.trim();
          if (tagsAllowed && !tagsAllowed.has(type as any)) {
            return Err({
              code: ErrorCode.UnrecognizedTag,
              message: `tag does not match any allowed tags`,
            });
          }
          break;
        }

        type += char;
        break;

      case State.Value:
        if (char === Token.Open) {
          return Err({
            code: ErrorCode.InvalidToken,
            message: `cannot use tag opening token '${
              Token.Open
            }' inside of tag`,
          });
        }

        // Finished consuming tag value, switch to close the tag.
        if (char === Token.Close) {
          state = State.Closing;
          value = value.trim();
          break;
        }

        value += char;
        break;

      case State.Closing:
        if (char !== Token.Close) {
          return Err({
            code: ErrorCode.UnclosedTag,
            message: `tag was not properly closed with two '${
              Token.Close
            }' tokens`,
          });
        }

        state = State.Text;

        tags.push({ type, value });
        // New text tag buffer
        type = TagType.Text;
        value = "";
        break;

      default:
        expectNever(state);
    }
  }

  if (state !== State.Text) {
    return Err({
      code: ErrorCode.UnclosedTag,
      message: "text parsed without closing a tag",
    });
  }

  // Flush the current tag.
  if (value.length) {
    tags.push({ type, value });
  }

  return Ok(tags);
}

function expectNever(_: never, msg = "unexpected code path"): never {
  throw new Error(msg);
}
