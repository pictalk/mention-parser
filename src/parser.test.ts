import * as Tags from "./parser";

const stringify = (x: any) => JSON.stringify(x, null, 2);

describe("Mention Parser", () => {
  it("should work", () => {
    let r = Tags.parse(`
Hello world! My best friend is {{mention:abcdefg123456}}, and together we love
{{hashtag:coding}}.
    `);

    expect(r).toBeDefined();
    expect(r.is_ok()).toBe(true);
    expect(r.unwrap()).toMatchSnapshot();
  });

  it("should work on tags with emoji types or values", () => {
    let text = `That moment when... {{hashtag:ðŸ’©}}`;
    let r = Tags.parse(text);

    let tokens = r.unwrap();
    let tag = tokens.find(t => t.type == "hashtag")!;
    expect(tag).toBeDefined();
    expect(tag.value).toBe("ðŸ’©");

    text = `That moment when... {{ðŸ’©:some-value}}`;
    r = Tags.parse(text);

    tokens = r.unwrap();
    tag = tokens.find(t => t.type == "ðŸ’©")!;
    expect(tag).toBeDefined();
    expect(tag.type).toBe("ðŸ’©");
  });

  it("should trim spaces in tag types and values", () => {
    let type = "foo";
    let value = "bar";
    let tests = [
      `Lorem ipsum dolor sit {{ ${type}:${value}}} amet consectetur adipisicing elit.`,
      `Lorem ipsum dolor sit {{ ${type} :${value}}} amet consectetur adipisicing elit.`,
      `Lorem ipsum dolor sit {{ ${type} : ${value}}} amet consectetur adipisicing elit.`,
      `Lorem ipsum dolor sit {{ ${type} : ${value} }} amet consectetur adipisicing elit.`,
      `Lorem ipsum dolor sit {{${type}: ${value}}} amet consectetur adipisicing elit.`,
      `Lorem ipsum dolor sit {{${type}: ${value} }} amet consectetur adipisicing elit.`,
      `Lorem ipsum dolor sit {{${type}:${value} }} amet consectetur adipisicing elit.`,
    ];

    for (let text of tests) {
      let tokens = Tags.parse(text).unwrap();
      let tag = tokens.find(t => t.type == type)!;
      expect(tag).toBeDefined();
      expect(tag.type).toBe(type);
      expect(tag.value).toBe(value);
    }
  });

  it("should work with spaces inside tag types and values", () => {
    let tests = [
      {
        type: "some type",
        value: "some value",
        text: `Now let's test {{ some type : some value }}.`,
      },
      {
        type: "another type with spaces",
        value: "a value with spaces",
        text: `Now let's test {{ another type with spaces : a value with spaces }}.`,
      },
    ];

    for (let { text, type, value } of tests) {
      let tokens = Tags.parse(text).unwrap();
      let tag = tokens.find(t => t.type == type)!;
      expect(tag).toBeDefined();
      expect(tag.type).toBe(type);
      expect(tag.value).toBe(value);
    }
  });

  it("should not trip up on curly brace tokens that aren't mustache tags", () => {
    let tests = [
      {
        type: "hashtag",
        value: "beach",
        text: JSON.stringify({
          user: {
            name: "Penelope",
            bio: "I love long walks on the {{hashtag:beach}}.",
          },
        }),
      },
      {
        type: "hashtag",
        value: "beach",
        text: "I love long walks on the {{{hashtag:beach}}}.",
      },
      {
        type: "hashtag",
        value: "beach",
        text: "I love long walks on the {{{{hashtag:beach}}}}.",
      },
      {
        type: "hashtag",
        value: "RealTalk",
        text: "{{{ hashtag : RealTalk }} I love long walks on the beach.",
        match: [
          { type: "text", value: "{" },
          { type: "hashtag", value: "RealTalk" },
          { type: "text", value: " I love long walks on the beach." },
        ],
      },
      {
        type: "hashtag",
        value: "RealTalk",
        text:
          "{{{ hashtag : RealTalk }}{{{{{mention: my husband  }}} and I love long walks on the beach.",
        match: [
          { type: "text", value: "{" },
          { type: "hashtag", value: "RealTalk" },
          { type: "text", value: "{{{" },
          { type: "mention", value: "my husband" },
          { type: "text", value: "} and I love long walks on the beach." },
        ],
      },
    ];

    for (let { text, type, value, match } of tests) {
      let r = Tags.parse(text);
      r.map_err(err => {
        console.error(text);
        console.error(err);
      });
      let tokens = r.unwrap();
      let tag = tokens.find(t => t.type == type)!;
      expect(tag).toBeDefined();
      expect(tag.type).toBe(type);
      expect(tag.value).toBe(value);
      if (match) {
        expect(tokens).toEqual(match);
      }
    }
  });

  it("should error when a tag is never closed", () => {
    let tests = [
      {
        label: "unclosed tag at end of string",
        text: `Let's burn this parser to the ground! {{hashtag:FuzzTesting}`,
      },
      {
        label: "unclosed tag in the middle of string",
        text: `Let's burn this {{hashtag:parser} to the ground!`,
      },
      {
        label: "invalid token inside tag",
        text: `Let's burn this {{hash{tag:parser}} to the ground!`,
      },
      {
        label: "invalid token inside tag",
        text: `Let's burn this {{hash}tag:parser}} to the ground!`,
      },
      {
        label: "tag closed before value",
        text: `Let's burn this {{hashtag}} to the ground!`,
      },
    ];

    for (let { text, label } of tests) {
      let r = Tags.parse(text);
      if (r.is_ok()) {
        console.error(`${label}:`, stringify(text));
        console.error(stringify(r));
      }
      expect(r.is_err()).toBe(true);
      let err = r.unwrap_err();

      // Ensure we provide a stable error shape
      expect(err.code).toBeDefined();
      expect(err.message).toBeDefined();
      // Ensure we provide a stable type
      expect(typeof err.message).toBe("string");
      expect(typeof err.code).toBe("number");
      // Ensure we know if our error codes change.
      // Consumers may rely on specific codes.
      expect(err.code).toMatchSnapshot(label);
    }
  });
});
