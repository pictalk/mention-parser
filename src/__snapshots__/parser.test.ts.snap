// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mention Parser should error when a tag is never closed: invalid token inside tag 1`] = `1`;

exports[`Mention Parser should error when a tag is never closed: invalid token inside tag 2`] = `1`;

exports[`Mention Parser should error when a tag is never closed: tag closed before value 1`] = `1`;

exports[`Mention Parser should error when a tag is never closed: unclosed tag at end of string 1`] = `2`;

exports[`Mention Parser should error when a tag is never closed: unclosed tag in the middle of string 1`] = `2`;

exports[`Mention Parser should work 1`] = `
Array [
  Object {
    "type": "text",
    "value": "
Hello world! My best friend is ",
  },
  Object {
    "type": "mention",
    "value": "abcdefg123456",
  },
  Object {
    "type": "text",
    "value": ", and together we love
",
  },
  Object {
    "type": "hashtag",
    "value": "coding",
  },
  Object {
    "type": "text",
    "value": ".
    ",
  },
]
`;

exports[`should error when tag does not match allowed tags case sensitive error for 'hashtag': case sensitive error for 'hashtag' 1`] = `3`;

exports[`should error when tag does not match allowed tags unknown tag 'hash': unknown tag 'hash' 1`] = `3`;
